modules/utxo-lib/src/bitgo/parseInput.ts

```diff
@@ -0,0 +1,418 @@
+  'scripthash',
+  'witnesspubkeyhash',
+  'witnessscripthash',
+  'taproot',
+  'taprootnofn',
+  'witnesscommitment',
+] as const;
+
+type InputType = typeof inputTypes[number];
+
+export function isPlaceholderSignature(v: number | Buffer): boolean {
+  if (Buffer.isBuffer(v)) {
+    return v.length === 0;
+  }
+  return v === 0;
+}
+
+export interface ParsedSignatureScript {
+  scriptType: ScriptType | 'p2pkh' | undefined;
+  isSegwitInput: boolean;
+  inputClassification: InputType;
+  p2shOutputClassification?: string;
+}
+
+export interface ParsedSignatureScriptUnknown extends ParsedSignatureScript {
+  scriptType: undefined;
+}
+
+export interface ParsedSignatureScriptP2PK extends ParsedSignatureScript {
+  scriptType: 'p2shP2pk';
+  inputClassification: 'scripthash';
+}
+
+export interface ParsedSignatureScriptP2PKH extends ParsedSignatureScript {
+  scriptType: 'p2pkh';
+  inputClassification: 'pubkeyhash';
+  signatures: [Buffer];
+  publicKeys: [Buffer];
+  pubScript?: Buffer;
+}
+
+export interface ParsedSignatureScript2Of3 extends ParsedSignatureScript {
+  scriptType: 'p2sh' | 'p2shP2wsh' | 'p2wsh';
+  inputClassification: 'scripthash' | 'witnessscripthash';
+  publicKeys: [Buffer, Buffer, Buffer];
+  signatures:
+    | [Buffer, Buffer] // fully-signed transactions with signatures
+    /* Partially signed transactions with placeholder signatures.
+       For p2sh, the placeholder is OP_0 (number 0) */
+    | [Buffer | 0, Buffer | 0, Buffer | 0];
+  pubScript: Buffer;
+}
+
+/**
+ * Keypath spends only have a single pubkey and single signature
+ */
+export interface ParsedSignatureScriptTaprootKeyPath extends ParsedSignatureScript {
+  scriptType: 'p2tr';
+  inputClassification: 'taproot';
+  publicKeys: [Buffer];
+  signatures: [Buffer];
+  pubScript: Buffer;
+}
+
+/**
+ * Taproot Scriptpath spends are more similar to regular p2ms spends and have two public keys and
+ * two signatures
+ */
+export interface ParsedSignatureScriptTaprootScriptPath extends ParsedSignatureScript {
+  scriptType: 'p2tr';
+  inputClassification: 'taproot';
+  publicKeys: [Buffer, Buffer];
+  signatures: [Buffer, Buffer];
+  controlBlock: Buffer;
+  /** Indicates the level inside the taptree. */
+  scriptPathLevel: number;
+  pubScript: Buffer;
+}
+
+export type ParsedSignatureScriptTaproot = ParsedSignatureScriptTaprootKeyPath | ParsedSignatureScriptTaprootScriptPath;
+
+type DecompiledScript = Array<Buffer | number>;
+
+/**
+ * Static script elements
+ */
+type ScriptPatternConstant =
+  | 'OP_0'
+  | 'OP_1'
+  | 'OP_2'
+  | 'OP_3'
+  | 'OP_CHECKMULTISIG'
+  | 'OP_CHECKSIG'
+  | 'OP_CHECKSIGVERIFY';
+
+/**
+ * Script elements that can be captured
+ */
+type ScriptPatternCapture = ':pubkey' | ':signature' | ':buffer' | { ':script': ScriptPatternElement[] };
+
+type ScriptPatternElement = ScriptPatternConstant | ScriptPatternCapture;
+
+/**
+ * Result for a successful script match
+ */
+type MatchResult = {
+  ':pubkey': Buffer[];
+  ':buffer': Buffer[];
```

Comment:
Since `:buffer` is only used for a controlblock, should we make the match a `:controlblock` that must be 33+32m in length?